pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java

```diff
@@ -181,6 +184,32 @@ protected void removeSegment(IndexSegment segment, MutableRoaringBitmap validDoc
     }
   }
 
+  /**
+   * When TTL is enabled for upsert, this function is used to remove expired keys from the primary key indexes.
+   *
+   * When committing consuming segment, we replaced the consuming segment with an immutable segments.
```

Comment:
Why do we want to do the cleanup of out-of-ttl keys during segment commit phase? This means the clean up is blocking and potentially slow down the ingestion. An alternative is to do it lazily as a background scheduled job. If I understand correctly, TTL for upsert table is an optimization, we do not need it synchronously. Here is the design doc about the choices

https://docs.google.com/document/d/1AhkZPkf4St3hj96IXyNs6NOiItxm3LXw3NAgNbG3oRo/edit?pli=1#heading=h.5ol3ncuq1lx2

Reply:
IIRC, we want to ensure the state for each replica to be the same. If we don't sync on the cleanup, each replica will potentially have different result, which can cause problem for upsert because during segment commit the segment can come from another server